client/state/themes/actions.js

```diff
@@ -106,13 +106,53 @@ export function receiveTheme( theme, siteId ) {
  * @return {Object}        Action object
  */
 export function receiveThemes( themes, siteId ) {
-	return {
-		type: THEMES_RECEIVE,
-		themes,
-		siteId
+	return ( dispatch, getState ) => {
+		const { filteredThemes } = filterThemes( getState, themes, siteId );
+		dispatch( {
+			type: THEMES_RECEIVE,
+			themes: filteredThemes,
+			siteId
+		} );
+	};
+}
+
+function receiveThemesQuery( themes, siteId, query, foundCount ) {
+	return ( dispatch, getState ) => {
+		const { filteredThemes, found } = filterThemes( getState, themes, siteId, query, foundCount );
+		dispatch( {
+			type: THEMES_REQUEST_SUCCESS,
+			themes: filteredThemes,
+			siteId,
+			query,
+			found,
+		} );
 	};
 }
 
+function filterThemes( getState, themes, siteId, query, found ) {
```

Comment:
Since this is a util function (and not e.g. an action thunk), and since it only uses `getState` once, I'd love if we could lose that and instead just pass, say, a `keepWpcom` flag...